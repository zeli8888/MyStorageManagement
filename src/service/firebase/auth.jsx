import {
    GoogleAuthProvider,
    GithubAuthProvider,
    OAuthProvider,
    signInWithPopup,
    setPersistence,
    browserSessionPersistence,
    signInWithEmailAndPassword,
    signOut,
    createUserWithEmailAndPassword,
    sendPasswordResetEmail,
} from 'firebase/auth';
import { firebaseAuth } from './firebaseConfig';

const googleProvider = new GoogleAuthProvider();
const githubProvider = new GithubAuthProvider();
const microsoftProvider = new OAuthProvider('microsoft.com');

microsoftProvider.setCustomParameters({
    prompt: 'select_account',
});

microsoftProvider.addScope('openid');
microsoftProvider.addScope('email');
microsoftProvider.addScope('profile');

// Sign in with Google functionality
export const signInWithGoogle = async () => {
    try {
        await setPersistence(firebaseAuth, browserSessionPersistence);
        const result = await signInWithPopup(firebaseAuth, googleProvider);
        return {
            success: true,
            user: result.user,
            error: null,
        };
    } catch (error) {
        return {
            success: false,
            user: null,
            error: error.message,
        };
    }
};

// Sign in with GitHub functionality
export const signInWithGithub = async () => {
    try {
        await setPersistence(firebaseAuth, browserSessionPersistence);
        const result = await signInWithPopup(firebaseAuth, githubProvider);
        return {
            success: true,
            user: result.user,
            error: null,
        };
    } catch (error) {
        return {
            success: false,
            user: null,
            error: error.message,
        };
    }
};

// Sign in with Microsoft functionality
export const signInWithMicrosoft = async () => {
    try {
        await setPersistence(firebaseAuth, browserSessionPersistence);
        const result = await signInWithPopup(firebaseAuth, microsoftProvider);
        return {
            success: true,
            user: result.user,
            error: null,
        };
    } catch (error) {
        return {
            success: false,
            user: null,
            error: error.message
        };
    }
};

// Sign in with email and password
export async function signInWithCredentials(email, password) {
    try {
        await setPersistence(firebaseAuth, browserSessionPersistence);
        const userCredential = await signInWithEmailAndPassword(
            firebaseAuth,
            email,
            password,
        );
        return {
            success: true,
            user: userCredential.user,
            error: null,
        };
    } catch (error) {
        return {
            success: false,
            user: null,
            error: error.message || 'Failed to sign in with email/password',
        };
    }
}

// Sign out functionality
export const firebaseSignOut = async () => {
    try {
        await signOut(firebaseAuth);
        return { success: true };
    } catch (error) {
        return {
            success: false,
            error: error.message,
        };
    }
};

// Sign up with email and password
export async function signUpWithCredentials(email, password) {
    try {
        const userCredential = await createUserWithEmailAndPassword(firebaseAuth, email, password);
        return { success: true, user: userCredential.user, error: null };
    } catch (error) {
        return { success: false, user: null, error: error instanceof Error ? error.message : "Registration failed" };
    }
}

// Send password reset email
export async function sendPasswordReset(email) {
    try {
        await sendPasswordResetEmail(firebaseAuth, email);
        return { success: true };
    } catch (error) {
        return { success: false, error: error instanceof Error ? error.message : "Password reset failed" };
    }
}

// Auth state observer
export const onAuthStateChanged = (callback) => {
    return firebaseAuth.onAuthStateChanged(callback);
};