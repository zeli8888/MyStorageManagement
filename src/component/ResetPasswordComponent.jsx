import React, { useState } from 'react';
import { useNavigate, Link } from 'react-router-dom';
import { TextField, Button, Box, Typography } from '@mui/material';
import { sendPasswordReset } from '../service/firebase/auth';

function ResetPasswordComponent() {
    const [email, setEmail] = useState('');
    const [message, setMessage] = useState('');
    const [error, setError] = useState('');
    const navigate = useNavigate();

    const handleSubmit = async (e) => {
        e.preventDefault();
        const result = await sendPasswordReset(email);

        if (result.success) {
            setMessage('Password reset email sent. Check your inbox.');
            setTimeout(() => navigate('/login'), 3000);
        } else {
            setError(result.error || 'Failed to send reset email');
        }
    };

    return (
        <Box sx={{
            maxWidth: { xs: '100%', sm: 400 },
            px: { xs: 2, sm: 0 },
            mx: 'auto',
            mt: { xs: 2, sm: 4 }
        }}>
            <Typography variant="h5" gutterBottom>Reset Password</Typography>
            <form onSubmit={handleSubmit}>
                <TextField
                    fullWidth
                    label="Email"
                    type="email"
                    value={email}
                    onChange={(e) => setEmail(e.target.value)}
                    margin="normal"
                    required
                />
                {message && <Typography color="success" sx={{ whiteSpace: 'normal' }}>{message}</Typography>}
                {error && <Typography color="error" sx={{ whiteSpace: 'normal' }}>{error}</Typography>}
                <Button
                    type="submit"
                    fullWidth
                    variant="contained"
                    sx={{
                        mt: { xs: 1, sm: 2 },
                        py: 1.5
                    }}
                >
                    Send Reset Email
                </Button>
                <Typography sx={{
                    mt: { xs: 1, sm: 2 },
                    fontSize: { xs: '0.875rem', sm: '1rem' }
                }}>
                    Remember your password? <Link to="/login">Sign In</Link>
                </Typography>
            </form>
        </Box>
    );
}

export default ResetPasswordComponent;
