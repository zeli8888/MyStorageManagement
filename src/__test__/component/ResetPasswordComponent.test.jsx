import { describe, it, expect, vi, afterEach, beforeEach } from 'vitest';
import { render, screen, cleanup, fireEvent } from '@testing-library/react';
import ResetPasswordComponent from '../../component/ResetPasswordComponent';
import { act } from 'react';
import { sendPasswordReset } from '../../service/firebase/auth';
import '@testing-library/jest-dom/vitest'

// Mock react-router-dom hooks and components
vi.mock('react-router-dom', () => ({
    useNavigate: vi.fn(),
    Link: ({ children, to }) => <a href={to}>{children}</a>,
}));

// Mock Firebase password reset service
vi.mock('../../service/firebase/auth', () => ({
    sendPasswordReset: vi.fn(),
}));

describe('ResetPasswordComponent', () => {
    beforeEach(() => {
        vi.clearAllMocks();
        sendPasswordReset.mockImplementation(() => vi.fn());
    })
    afterEach(() => {
        cleanup();
    });

    it('should match snapshot', () => {
        const { container } = render(<ResetPasswordComponent />);
        expect(container).toMatchSnapshot();
    });

    it('should display success message and navigate after form submission', async () => {
        sendPasswordReset.mockResolvedValue({ success: true });

        render(<ResetPasswordComponent />);

        act(() => {
            fireEvent.change(screen.getByLabelText(/Email/i), { target: { value: 'test@example.com' } });
            fireEvent.click(screen.getByText(/Send Reset Email/i));
        })

        await screen.findByText(/Password reset email sent. Check your inbox./i);
        expect(sendPasswordReset).toHaveBeenCalledWith('test@example.com');
    });

    it('should display error message when password reset fails', async () => {
        sendPasswordReset.mockResolvedValue({ success: false, error: 'User not found' });

        render(<ResetPasswordComponent />);

        act(() => {
            fireEvent.change(screen.getByLabelText(/Email/i), { target: { value: 'invalid@example.com' } });
            fireEvent.click(screen.getByText('Send Reset Email'));
        })

        await screen.findByText('User not found');
        expect(sendPasswordReset).toHaveBeenCalledWith('invalid@example.com');
    });

    it('should display default error message when password reset fails without error message', async () => {
        sendPasswordReset.mockResolvedValue({ success: false });

        render(<ResetPasswordComponent />);

        act(() => {
            fireEvent.change(screen.getByLabelText(/Email/i), { target: { value: 'invalid@example.com' } });
            fireEvent.click(screen.getByText('Send Reset Email'));
        })

        await screen.findByText(/Failed to send reset email/i);
        expect(sendPasswordReset).toHaveBeenCalledWith('invalid@example.com');
    });

    it('should show login link', () => {
        render(<ResetPasswordComponent />);
        expect(screen.getByText(/Remember your password/i)).toBeInTheDocument();
        expect(screen.getByText(/Sign In/i).closest('a')).toHaveAttribute('href', '/login');
    });

    it('should validate email input', async () => {
        render(<ResetPasswordComponent />);
        const emailInput = screen.getByLabelText(/Email/i);

        // Test empty email validation
        act(() => {
            fireEvent.change(emailInput, { target: { value: '' } });
            fireEvent.click(screen.getByText('Send Reset Email'));
        })
        expect(emailInput).toBeInvalid();
    });
});
