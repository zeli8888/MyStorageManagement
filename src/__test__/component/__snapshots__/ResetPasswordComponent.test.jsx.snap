// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResetPasswordComponent > should match snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-1uzsq72"
  >
    <h5
      class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1f9xtph-MuiTypography-root"
    >
      Reset Password
    </h5>
    <form>
      <div
        class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-16bevx5-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r0:"
          id=":r0:-label"
        >
          Email
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-1ljffdk-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1blp12k-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="email"
            class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            required=""
            type="email"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-1n64csd-MuiNotchedOutlined-root"
            >
              <span>
                Email
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth css-i7we6i-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Send Reset Email
      </button>
      <p
        class="MuiTypography-root MuiTypography-body1 css-pw9lyv-MuiTypography-root"
      >
        Remember your password? 
        <a
          href="/login"
        >
          Sign In
        </a>
      </p>
    </form>
  </div>
</div>
`;
